### ISSUE TITLE
Add triangular matrix determinant formula

### ISSUE THREAD

**Issue by sylee957** [sympy/sympy#18607]
Why not adding a determinant formula for triangular matrix, with `is_upper` or `is_lower` check such that it does not attempt to compute in expensive way.
We have triangular matrix eigenvalue formula at least
https://proofwiki.org/wiki/Determinant_of_Triangular_Matrix

```python3
>>> from sympy import *

>>> n = 20
>>> A = MatrixSymbol('A', n, n)
>>> A = Matrix(n, n, lambda i, j: A[i, j] if i <= j else S.Zero)
>>> A.det()
```

Same thing can apply for `charpoly`

**Comment by sudoWin** [comment]
Hi @sylee957 , I would like to work and contribute to this issue. I have decent experience in Python and Sympy usage, and over the last week I have made myself familiar with the code base of Sympy and Symengine.

**Comment by sylee957** [comment]
Okay, you can start working on by referencing how `eigenvals` handle the trivial case.
I'd also expect to add some test cases about it.

**Comment by sudoWin** [comment]
Hi,
I need some help.
Till now I have 
1.Finished the changes and added is_upper is_lower check in _det and _charpoly, 
2.Updated docstrings in the local copy
3.Added some new tests
4.Ran all the tests by ./bin/test , ./bin/test quality, and flake8
5.Committed all the changes to local branch.

However for last few hours I have been trying hard to push the changes to my github fork, but I am failing in that step,It would be great if someone could help.

Also, I had one more query. It looks like that sympy:master has been updated since I forked it last time.How can I make my fork even with sympy:master?

Thanks
